compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64RedundantInstructionElimination.java

```diff
@@ -0,0 +1,190 @@
+ */
+public final class AArch64RedundantInstructionElimination extends PreAllocationOptimizationPhase {
+
+    @Override
+    protected void run(TargetDescription target, LIRGenerationResult lirGenRes, PreAllocationOptimizationContext context) {
+        new Optimization(lirGenRes.getLIR()).apply();
+    }
+
+    private static final class Optimization {
+        private final LIR lir;
+        private final VariableMap<DefEntry> unUsedDefs;
+        private final DebugContext debug;
+
+        private Optimization(LIR lir) {
+            this.lir = lir;
+            this.debug = lir.getDebug();
+            this.unUsedDefs = new VariableMap<>();
+        }
+
+        @SuppressWarnings("try")
+        private void apply() {
+            try (Indent indent = debug.logAndIndent("AArch64RedundantInstructionElimination")) {
+                try (DebugContext.Scope s = debug.scope("CollectUnusedInstructions")) {
+                    for (AbstractBlockBase<?> b : lir.getControlFlowGraph().getBlocks()) {
+                        collectUnusedInBlock(b);
+                    }
+
+                    if (!unUsedDefs.isEmpty()) {
+                        debug.log("Redundant instructions that will be removed later: %s", unUsedDefs);
+                    }
+                } catch (Throwable e) {
+                    throw debug.handle(e);
+                }
+
+                try (DebugContext.Scope s = debug.scope("DeleteUnusedInstructions")) {
+                    unUsedDefs.forEach(this::deleteInstruction);
+                    // Remove unused instructions for each block.
+                    for (AbstractBlockBase<?> b : lir.getControlFlowGraph().getBlocks()) {
+                        removeUnusedInBlock(b);
+                    }
+                } catch (Throwable e) {
+                    throw debug.handle(e);
+                }
+            }
+        }
+
+        /**
+         * Collect unused instructions for a {@code block}.
+         */
+        private void collectUnusedInBlock(AbstractBlockBase<?> block) {
+            InstructionValueConsumer defConsumer = (instruction, value, mode, flags) -> {
+                if (isVariable(value) && isEligibleOp(instruction)) {
+                    DefEntry def = new DefEntry(block, instruction, value);
+                    unUsedDefs.put(((Variable) value), def);
+                }
+            };
+
+            InstructionValueConsumer useConsumer = (instruction, value, mode, flags) -> {
+                if (isVariable(value)) {
+                    DefEntry def = unUsedDefs.get((Variable) value);
+                    if (def != null) {
+                        assert def.value.equals(value);
+                        // Remove the instruction that defined this value from the unused map.
+                        LIRInstruction inst = def.instruction;
+                        unUsedDefs.filter(d -> (!d.instruction.equals(inst)));
```

Comment:
Since we are in SSA form, there can be only one instruction defining a variable. Thus, I think the code can be changed to `unUsedDefs.remove((asVariable(value))`.

Reply:
Actually, directly calling `remove` is all right for `ConstantLoadOp`. Thanks! My original concern is the `Def` with an array type like `@Def({REG, STACK}) private Value[] incomingValues;` of `LabelOp`. If an instruction has a definition with an array type. It actually has multiply definition values. And if any of the member is used, the instruction should not be removed. So I use the `filter` instead of "remove".  Anyway it seems this might not be possible for all the existed `ConstantLoadOp`.

Reply:
> If an instruction has a definition with an array type. It actually has multiply definition values.

That is right. In that case, there is still no need to use `filter` (which can be expensive I think) but do something like this (more or less pseudocode):
```java
InstructionValueConsumer removeConsumer = (instruction, value, mode, flags) -> {
    if (isVariable(value)) {
        unUsedDefs.remove(asVariable(value));
    }
};
...
inst.visitEachOutput(removeConsumer);
```

Reply:
OK, it seems reasonable! So as a conclusion, which do you think is better:  add these codes to handle the corner case, or just a directly `unUsedDefs.remove(asVariable(value))` is enough? I think the latter is ok. Your opinion? Thanks!

Reply:
I agree. However, I'd add a comment why this is OK and an assertion that `inst` in no longer in any entry of the `unUsedDefs` map to catch error early if our assumption changes.

Reply:
Good idea! I will update the patch soon! Thank you!